### ISSUE TITLE
Default to required field

### ISSUE THREAD

**Issue by cereallarceny** [saiichihashimoto/sanity-typed#222]
When using the former library `schema-builder`, something I took for granted was that every field was automatically required by default. As part of the migration process to the `types` library (which I still much prefer), I've lost all my required fields. Is there some easy way to automatically default to fields being default, with the ability to include an `optional: true` property on the field? Perhaps we could pass some options into `defineConfig` to make this happen, something like this:

```ts
const config = defineConfig(
  [
    {
      projectId,
      schema,
      plugins,
      title: 'Production',
      name: 'production',
      basePath: '/prod',
      dataset: 'production',
      icon: RocketIcon,
    },
    {
      projectId,
      schema,
      plugins,
      title: 'Development',
      name: 'development',
      basePath: '/dev',
      dataset: 'development',
      icon: RobotIcon,
    },
  ],
  {
    requiredByDefault: true
  }
);
```

While I think the idea behind this library is "0 barrier to entry when compared with the default sanity `defineX` methods", this would be a seemingly unobtrusive addition that would offer great benefit to type safety - which is the whole point.

Thoughts @saiichihashimoto ?

**Comment by saiichihashimoto** [comment]
This library's goal is to add types to sanity's outputs (config, groq, etc) while changing the API as little as possible. I agree that required by default _would_ be ideal (would also be what I prefer), but QOL improvements isn't the objective here, especially when it deviates from the API. With `schema-builder` I was getting a bunch of comments about how required/optional worked, since it was deviating in this exact way.

I think this is a suggestion better made to [sanity](https://github.com/sanity-io/sanity).

This isn't complete, but another suggestion is to do something like this:

```typescript
import { defineField as defineFieldOptional } from "@sanity-typed/types";

const defineField = (args: ProbablyNeedsSomeHelpToBeTypedProperly) => defineFieldOptional({
  validation: (Rule) => Rule.required(),
  ...args,
});

// use defineField(...) normally.
```

Something like that.

**Comment by cereallarceny** [comment]
So @saiichihashimoto, I totally get the idea of what you're talking about. Interestingly enough, I already tried what you're suggesting (overriding `defineField` to be required and creating a `defineFieldOptional`). But the thing that stopped me was the types...

If you're not planning on supporting QOL functionality - would you at least be able to provide the ability for developers to have ready-to-go types and allow us to extend the default functionality. If I wanted to create a custom version of `defineField`, how would one go about doing this? Could you provide something more specific than `ProbablyNeedsSomeHelpToBeTypedProperly`?

**Comment by saiichihashimoto** [comment]
Yeah I get what you're saying. I think the issue you'd have typing this is that any typing of that new `defineField` is going to redefine and therefore tie itself to the implementation details (ie the generics) of the sanity-typed `defineField`. Looking into it further, I think the easiest way around this is something like this (which I've tested locally):

```typescript
// All you're enforcing is that `validation` has a `required` method, and this returns whatever it's return type is.
const validation = <T>(Rule: { required: () => T }) => Rule.required();

const foo = defineField({
  name: "foo",
  type: "object",
  fields: [
    defineField({
      name: "bar",
      type: "boolean",
      // ... so this will infer properly, regardless of implementation details.
      validation,
    }),
  ],
});

expectType<_InferValue<typeof foo>>().toStrictEqual<{
  bar: boolean;
}>;
```

It's also nice that this is composition, so you can easily mix in other common things you want to set. You can find ways to accomplish what you were aiming for, but I don't think they're good ideas, considering they tie you to implementation details I'm likely to change and not consider a breaking change.

I'm thinking out loud here, but I think there's a few points of clarity that are good for the future:

- The goal is to provide typescript types for the content lake documents (and how they end up via the GROQ and client)
- It doesn't have any (much) runtime changes.
  - This is a better way to describe it than "changing the API as little as possible" or "QOL improvements". If you look at the exported `.js` files, they purely re-export the sanity exports.
  - The only exception will be to `createClient` in the upcoming release from `alpha`, because of a really obnoxious unavoidable nuance of typescript generics.
- It doesn't have any (much) typing requirements, ie providing generics or retyping anything. Besides the unavoidable step of `InferSchemaValues`, there isn't anywhere that we're requiring you to provide types anywhere.
  - Of course, once we include `groq` and `client`, it'll need you to provide the types from your config, but that'll actually be intentional.
  - Unfortunately, there's also places that `as const` will be needed which makes me very sad.

**Comment by saiichihashimoto** [comment]
Oh I forgot to add: please create an issue in the sanity repo and feel free to link to this issue. I would also like a native way to have everything be required by default. I understand why a CMS would have everything optional be default (thinking of the sanity user's experience, it's nice to just fill in whatever information you happen to have) but using rule validation for this is kind of unpleasant.

**Comment by cereallarceny** [comment]
Understood, I can roll with that for now. I'll post an example if I get something better working.
